### PR TITLE
make-based: Add support for app-elfloader

### PR DESCRIPTION
**Pull request by razvand**
Add configuration files and `do.sh` script to setup, build and run `app-elfloader`.

### CODE REVIEW

File: make-based/app-elfloader/files/.config [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+# CONFIG_LIBSYSCALL_SHIM_LIBCSTUBS is not set
+CONFIG_LIBSYSCALL_SHIM_HANDLER=y
+CONFIG_LIBSYSCALL_SHIM_HANDLER_ULTLS=y
+
+#
+# Debugging
+#
+# CONFIG_LIBSYSCALL_SHIM_DEBUG_SYSCALLS is not set
+# CONFIG_LIBSYSCALL_SHIM_DEBUG_HANDLER is not set
+CONFIG_LIBSYSCALL_SHIM_STRACE=y
```
Comment (StefanJum) [inline make-based/app-elfloader/files/.config:132]: Should we enable this in the non-debug `.config` file?
I think it gives a bit too much output, I would not use it in the non-debug configuration.

**Review by StefanJum — COMMENTED** [review]
Everything works fine, see the one comment.

File: make-based/app-elfloader/files/.config (hunk 1 above)
Comment (razvand) [reply make-based/app-elfloader/files/.config:132]: We can create three configuration files: plain, strace, and full debug. It's very important to have this one for debugging. The debug one is too verbose.

File: make-based/app-elfloader/do.sh [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    git rebase csvancea/fix/argv-parser
+    git rebase csvancea/fix/unaligned-stack
+
+    if test ! -d ../run-app-elfloader; then
+        git clone https://github.com/unikraft/run-app-elfloader ../run-app-elfloader
+    fi
+
+    if test ! -d ../static-pie-apps; then
+        git clone https://github.com/unikraft/static-pie-apps ../static-pie-apps
+    fi
```
Comment (StefanJum) [inline make-based/app-elfloader/do.sh:46]: I think the [`dynamic-apps` repository](https://github.com/unikraft/dynamic-apps) should also be listed here.
Running anything not static gives `failed to open '../dynamic-apps/[...]': No such file or directory`

**Review by StefanJum — CHANGES_REQUESTED** [review]
See the one comment

File: make-based/app-elfloader/do.sh (hunk 2 above)
Comment (razvand) [reply make-based/app-elfloader/do.sh:46]: Indeed, will fix.

**Review by StefanJum — COMMENTED** [review]
Looks good @razvand, but when I try to run any dynamically linked application I get this error:

```
[    0.262278] CRIT: [libkvmplat] <trace.c @   41> RIP: 0000000000000000 CS: 0008
[    0.264376] CRIT: [libkvmplat] <trace.c @   42> RSP: 000000047f92fb08 SS: 0010 EFLAGS: 00010202
[    0.267032] CRIT: [libkvmplat] <trace.c @   44> RAX: 0000000000000000 RBX: 0000000000000200 RCX: 0000000000000000
[    0.269951] CRIT: [libkvmplat] <trace.c @   46> RDX: 000000047f92fb80 RSI: 000000047f92fb90 RDI: 000000040006f968
[    0.272968] CRIT: [libkvmplat] <trace.c @   48> RBP: 0000001000295670 R08: 000000047f92fba0 R09: 00000010005f0d80
[    0.275996] CRIT: [libkvmplat] <trace.c @   50> R10: 000000040006ef68 R11: 0000000000000246 R12: 0000000000000000
[    0.279007] CRIT: [libkvmplat] <trace.c @   52> R13: 000000047f92fb90 R14: 000000047f92fb80 R15: 000000040006f060
[    0.281920] CRIT: [libkvmplat] <trace.c @   86> base is 0x1000295670 caller is 0x401f0fc3f30678
[    0.284401] CRIT: [libkvmplat] <trace.c @   86> base is 0xc08500359bee058b Unhandled Trap 13 (general protection), error code=0x0
[    0.287617] CRIT: [libkvmplat] <trace.c @   41> RIP: 0000000000104071 CS: 0008
[    0.289569] CRIT: [libkvmplat] <trace.c @   42> RSP: 00000000001efee0 SS: 0010 EFLAGS: 00010002
[    0.291910] CRIT: [libkvmplat] <trace.c @   44> RAX: 00000000001efdf8 RBX: c08500359bee058b RCX: 0000000000000018
[    0.294727] CRIT: [libkvmplat] <trace.c @   46> RDX: 0000003000000030 RSI: 0000000000000000 RDI: 00000000001a1f2a
[    0.297669] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000001efef0 R08: 0000000000000046 R09: 00000000000000ff
[    0.300516] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000002 R11: 0000000000000000 R12: 0000000000000000
[    0.303476] CRIT: [libkvmplat] <trace.c @   52> R13: 00000000001eff50 R14: 000000047f92fb80 R15: 000000040006f060
```

Is it something that is supposed to happed? If not I can jump into debugging it.

**Comment by razvand** [comment]
> Looks good @razvand, but when I try to run any dynamically linked application I get this error:

Are you sure about that? It doesn't replicate on my side. Did you try running [the pre-built images of `app-elfloader`](https://github.com/unikraft/run-app-elfloader)? Maybe it's a compiler / build environment issue.

**Comment by StefanJum** [comment]
> Are you sure about that? It doesn't replicate on my side. Did you try running [the pre-built images of `app-elfloader`](https://github.com/unikraft/run-app-elfloader)? Maybe it's a compiler / build environment issue.

Yes, the issue appears only with the pre-build images (running `./run_app.sh helloworld`). If I build `app-elfloader` from scratch, then everything works fine.
It also only occurs with dynamically linked executables, the static-pie ones work fine with both the pre-build and the manual build elfloader.

It seems it breaks when [this commit](https://github.com/unikraft/app-elfloader/pull/13/commits/a4d823bb783453e598cf2f228afa26abb1430e58) is added, which seems a bit weird. I'll keep looking into it, try testing it on another machine too, in the meantime we can merge this if the issue is not reproducible for you.

**Pull request merged**